File: 15-3sum/3sum.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function(nums) {
    nums.sort((a,b)=>a-b)
    let res =[]
    for(let i=0;i<nums.length;i++ ){
    if (i > 0 && nums[i] === nums[i - 1]) continue;
    let l =i+1
    let r =nums.length-1
    let target = nums[i]
    while(l<r){
        let t = nums[l]+nums[r]
        if(t===-target){
            res.push([nums[i],nums[l],nums[r]])
          
        while(l<r&&nums[l]===nums[l+1]) l++
        while(l<r && nums[r]===nums[r-1]) r--

        l++
        r--
        }
        else if(t<-target){
            l++
        }
        else{
            r--
        }
    }
    }
    return res
};

File: 15-3sum/3sum.py
class Solution(object):
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        freq = Counter(nums)
        print(freq)
        nums.sort()
        ans = []
        for k in range(len(nums)):
            if k>0 and nums[k] == nums[k-1]: continue

            i=k+1
            j=len(nums)-1
            while i<j:
                t = nums[k]+nums[i]+nums[j]
                if t<0:
                    i+=1
                elif t>0: j-=1
                else:
                    ans.append([nums[k],nums[i],nums[j]])
                    i+=1
                    j-=1
                    while i<j and nums[i]==nums[i-1]:i+=1
                    while i<j and nums[j] == nums[j+1]:j-=1
        return ans


        